File: 백준/Gold/4179. 불！/불！.java
import java.io.*;
import java.util.*;

/**
 * 문제: 미로 탈출
 * 조건: 장애물의 퍼짐
 */

class Main {

    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    static int n;
    static int m;
    static char[][] board = new char[1002][1002];
    static int[][] fd = new int[1002][1002];
    static int[][] jd = new int[1002][1002];
    static Queue<int[]> FQ = new ArrayDeque<>();
    static Queue<int[]> JQ = new ArrayDeque<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        
        for (int i = 0; i < n; i++) {
            String input = br.readLine();
            for (int j = 0; j < m; j++) {
                board[i][j] = input.charAt(j);
                fd[i][j] = -1;
                jd[i][j] = -1;
                if (board[i][j] == 'F') {
                    FQ.offer(new int[]{i, j});
                    fd[i][j] = 0;
                }
                else if (board[i][j] == 'J') {
                    JQ.offer(new int[]{i, j});
                    jd[i][j] = 0;
                }
            }
        }
        int result = solve();
        System.out.print(result != -1 ? result : "IMPOSSIBLE");
    }

    private static int solve() {
        fireBFS();
        while (!JQ.isEmpty()) {
            var cur = JQ.poll();
            int nextDist = jd[cur[0]][cur[1]] + 1;
            if (cur[0] == 0 || cur[0] == n - 1 || cur[1] == 0 || cur[1] == m - 1) {
                return nextDist;
            }

            for (int i = 0; i < 4; i++) {
                int nx = cur[0] + dx[i];
                int ny = cur[1] + dy[i];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (board[nx][ny] == '#' || jd[nx][ny] != -1) continue;
                if (fd[nx][ny] != -1 && nextDist >= fd[nx][ny]) continue;
                jd[nx][ny] = nextDist;
                JQ.offer(new int[]{nx, ny});
            }
        }
        return -1;
    }

    private static void fireBFS() {
        while (!FQ.isEmpty()) {
            var cur = FQ.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cur[0] + dx[i];
                int ny = cur[1] + dy[i];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (board[nx][ny] == '#' || fd[nx][ny] != -1) continue;
                fd[nx][ny] = fd[cur[0]][cur[1]] + 1;
                FQ.offer(new int[]{nx, ny});
            }
        }
    }
}

File: 백준/Gold/4179. 불！/불！.kt
import java.util.*

val dx = intArrayOf(1, 0, -1, 0);
val dy = intArrayOf(0, 1, 0, -1);

fun main() = with(System.`in`.bufferedReader()) {
    val (r, c) = readLine().split(" ").map { it.toInt() }
    val board = Array(r) { readLine().toCharArray() }

    val df = Array(r) { IntArray(c) { -1 } }
    val dp = Array(r) { IntArray(c) { -1 } }

    val fq = ArrayDeque<Pair<Int, Int>>()
    val pq = ArrayDeque<Pair<Int, Int>>()
    for (i in 0 until r) {
        for (j in 0 until c) {
            if (board[i][j] == 'F') {
                fq.offer(i to j)
                df[i][j] = 0
            } else if (board[i][j] == 'J') {
                pq.offer(i to j)
                dp[i][j] = 0

                if (i == 0 || i == r - 1 || j == 0 || j == c - 1) {
                    print(1)
                    return@with
                }
            }
        }
    }

    while (fq.isNotEmpty()) {
        val (cx, cy) = fq.poll()

        for (i in 0 until 4) {
            val nx = cx + dx[i]
            val ny = cy + dy[i]

            if (nx !in 0..r-1 || ny !in 0..c-1) continue
            if (board[nx][ny] == '#') continue
            if (df[nx][ny] != -1) continue
            df[nx][ny] = df[cx][cy] + 1
            fq.offer(nx to ny)
        }
    }

    while (pq.isNotEmpty()) {
        val (cx, cy) = pq.poll()

        for (i in 0 until 4) {
            val nx = cx + dx[i]
            val ny = cy + dy[i]

            if (nx !in 0..r-1 || ny !in 0..c-1) continue
            if (board[nx][ny] == '#') continue
            if (dp[nx][ny] != -1) continue
            if (df[nx][ny] == -1 || df[nx][ny] > dp[cx][cy] + 1) {
                dp[nx][ny] = dp[cx][cy] + 1
                if (nx == 0 || nx == r - 1 || ny == 0 || ny == c - 1) {
                    print(dp[nx][ny] + 1)
                    return@with
                }
                pq.offer(nx to ny)
            }
        }
    }

    print("IMPOSSIBLE")
}